python/GafferSceneUI/ShaderUI.py

```diff
@@ -621,6 +621,131 @@ def __parameterValues( self ) :
 
 		return None
 
+# A path filter that keeps a path if it, any of its ancestors or any of its descendants have a
+# property `propertyName` that matches `patterns`.
+
+class _PathMatcherPathFilter( Gaffer.PathFilter ) :
+
+	def __init__( self, patterns, rootPath = None, propertyName = "name", userData = {} ) :
+
+		Gaffer.PathFilter.__init__( self, userData )
+
+		self.__patterns = patterns
+		self.__rootPath = rootPath
+		self.__propertyName = propertyName
+
+		self.__pathMatcherDirty = True
+		self.__pathMatcher = IECore.PathMatcher()
+
+		self.setRootPath( rootPath )
+
+	def setMatchPatterns( self, patterns ) :
+
+		if self.__patterns == patterns :
+			return
+
+		self.__patterns = patterns
+		self.__pathMatcherDirty = True
+		self.changedSignal()( self )
+
+	def getMatchPatterns( self ) :
+
+		return self.__patterns
+
+	def setRootPath( self, rootPath ) :
+
+		if rootPath is None :
+			self.__rootPathChangedConnection = None
+			self.__rootPath = rootPath
+			self.__paths = []
+			self.__pathMatcherDirty = True
+			return
+
+		if self.__rootPath == rootPath :
+			return
+
+		self.__rootPath = rootPath
+		self.__updatePaths()
+
+		self.__rootPathChangedConnection = self.__rootPath.pathChangedSignal().connect(
+			Gaffer.WeakMethod( self.__rootPathChanged ),
+			scoped = True
+		)
+		self.__pathMatcherDirty = True
+
+	def getRootPath( self ) :
+
+		return self.__rootPath
+
+	def setPropertyName( self, propertyName ) :
+
+		if( self.__propertyName == propertyName ) :
+			return
+
+		self.__propertyName = propertyName
+		self.__pathMatcherDirty = True
+		self.changedSignal()( self )
+
+	def getPropertyName( self ) :
+
+		return self.__propertyName
+
+	def _filter( self, paths, canceller ) :
+
+		if len( paths ) == 0 :
+			return []
+
+		self.__updatePathMatcher()
+
+		if self.__pathMatcher.isEmpty() :
+			return []
+
+		result = [ p for p in paths if self.__pathMatcher.match( str( p ) ) ]
+
+		return result
+
+	def __updatePathMatcher( self ) :
+
+		if not self.__pathMatcherDirty or len( self.__patterns ) == 0 or self.__propertyName == "" :
+			return
+
+		newPathMatcher = IECore.PathMatcher()
+
+		for p in self.__paths :
+			property = p.property( self.__propertyName )
+			if property is None :
+				continue
+
+			for pattern in self.__patterns :
+				if IECore.StringAlgo.match( property, pattern ) :
+					newPathMatcher.addPath( str( p ) )
+					break
+
+		if self.__pathMatcher == newPathMatcher :
+			return
+
+		self.__pathMatcher = newPathMatcher
+		self.changedSignal()( self )
+
+	def __updatePaths( self ) :
+
+		def collectPaths( parentPath ) :
+			result = []
+			for p in parentPath.children() :
+				result.append( p )
+				result += collectPaths( p )
+
+			return result
+
+		self.__paths = collectPaths( self.__rootPath )
+
+	def __rootPathChanged( self, path ) :
+
+		self.__updatePaths()
```

Comment:
This is going to rebuild the paths immediately any time `pathChanged()` is signalled. Why don't we just set `__pathMatcherDirty` and iterate the paths lazily in `__updatePathMatcher`? In other words, why do we even need `self.__paths`?

Reply:
That is a reasonable argument for sure. I guess maybe I'd counter by saying that storing all the Path objects could be quite expensive in the general case, vs just storing the matches in a PathMatcher. And I quite liked the simplicity of removing the additional state. All a bit hypothetical given the current use case, I know. I'll defer to your judgment - certainly if it makes a tangible performance difference to the current use case, then storing `__paths` is worthwhile.

Reply:
I'll give both a try and see if I can pick up on a performance difference.